### ISSUE TITLE
STBI: 5:6:5:0 BMP doesn't load correctly

### ISSUE THREAD

**Issue by 8infy** [nothings/stb#870]
Long story short I'm making my own image-loading library and I use stbi as a unit test reference. So far every single image (all bmps) i've tried have been working great. However, this one image seems to be breaking stbi.

Here's a simple 4x4 image. I have it as a 5:6:5:0 BMP.
![image](https://user-images.githubusercontent.com/34326714/71825601-3f6ffa00-30ad-11ea-8df9-8bce0e5ae196.png)
When I try to load it with stbi it loads something completely off. Heres what the buffer looks like after loading (0-6 also zeroes): 
![image](https://user-images.githubusercontent.com/34326714/71825690-6a5a4e00-30ad-11ea-9181-cce2113d4915.png)

And heres what it should look like (last number is 33):
![image](https://user-images.githubusercontent.com/34326714/71826016-0be19f80-30ae-11ea-8606-c170ce5b1151.png)

I can send you the image if you need it, but any 5:6:5:0 BMP should work afaik.

PS: I use [this](https://online-converting.com/image/convert2bmp/) website for BMP conversions

**Comment by nothings** [comment]
that's not a 4x4 image

**Comment by 8infy** [comment]
This is the screenshot of the image

> that's not a 4x4 image

This is a screenshot of the actual image 😄 
Github doesn't support this format.

**Comment by nothings** [comment]
you gotta post the actual image that's breaking. i don't care what it looks like, just the actual failing bytes.

**Comment by 8infy** [comment]
Try this:
[t656.zip](https://github.com/nothings/stb/files/4026600/t656.zip)

**Comment by nothings** [comment]
I don't know. I stepped through in the debugger and watched it decode the first pixel as white (255,255,255), which is the bottom left corner of the image and should be black.

The file format must be different than the code thinks, but I don't know how or in what way, as everything seemed reasonable in what I saw.

**Comment by 8infy** [comment]
> I don't know. I stepped through in the debugger and watched it decode the first pixel as white (255,255,255), which is the bottom left corner of the image and should be black.
> 
> The file format must be different than the code thinks, but I don't know how or in what way, as everything seemed reasonable in what I saw in the debugger.
> 
> I also verified that it opened correctly in mspaint, i.e. differently from what stb_image does.
Yup. Probably misreading some header data somewhere.

**Comment by nothings** [comment]
Ok, I figured it out. Fixed for next version. I don't know when that will be.

The bug: BMP contains an offset to the pixel data, and after parsing the header, stb_image attempts to seek to that offset. (Normally, this does nothing as the pixel data is immediately after the header). However, it does that not by looking at file location, but by skipping the intervening data. That is, it subtracts the amount of header data it read from the offset of the pixel data, and skips that much data, which should put it at the offset.

But, in the case where the header contains bitmasks indicating how to decode r/g/b, those bitmasks weren't accounted for in the how-much-header-data-it-saw (there's an explicit "header size" in the file and that size doesn't include the bitmasks), so when it attempted to skip to the pixel data, it wouldn't realize the file pointer was already 12 bytes ahead of where it thought, so it would skip 12 bytes too far and start decoding from there, in this case missing the first 6 pixels.
